src/d3a/models/market/market_validators.py

```diff
@@ -0,0 +1,106 @@
+"""Validators for d3a markets."""
+
+from abc import ABC, abstractmethod
+from typing import Dict
+
+from d3a.models.market import Offer, Bid
+
+
+class Requirement(ABC):
+    """Define an offer/bid requirement with is_satisfied interface."""
+
+    @classmethod
+    @abstractmethod
+    def is_satisfied(cls, offer: Offer, bid: Bid, requirement: Dict) -> bool:
+        """Check whether a requirement is satisfied."""
+
+
+class PreferredPartnersRequirement(Requirement):
+    """Check if preferred_trading_partners requirement is satisfied for both bid and offer."""
+
+    @classmethod
+    def is_satisfied(cls, offer: Offer, bid: Bid, requirement: Dict) -> bool:
+        preferred_trading_partners = requirement.get("preferred_trading_partners", [])
+        assert isinstance(preferred_trading_partners, list),\
+            f"Invalid requirement {requirement}"
+        if preferred_trading_partners and not (
+                bid.buyer_id in preferred_trading_partners or
+                bid.buyer_origin_id in preferred_trading_partners or
+                offer.seller_id in preferred_trading_partners or
+                offer.seller_origin_id in preferred_trading_partners):
+            return False
+        return True
+
+
+class EnergyTypeRequirement(Requirement):
+    """Check if energy_type requirement of bid is satisfied."""
+
+    @classmethod
+    def is_satisfied(cls, offer: Offer, bid: Bid, requirement: Dict) -> bool:
+        bid_required_energy_types = requirement.get("energy_type", [])
+        assert isinstance(bid_required_energy_types, list), \
+            f"Invalid requirement {requirement}"
+        offer_energy_type = (
+                offer.attributes or {}).get("energy_type", None)
+        if bid_required_energy_types and offer_energy_type not in bid_required_energy_types:
+            return False
+        return True
+
+
+class HashedIdentityRequirement(Requirement):
+    """Check if hashed_identity requirement of bid is satisfied."""
+
+    @classmethod
+    def is_satisfied(cls, offer: Offer, bid: Bid, requirement: Dict) -> bool:
+        bid_required_hashed_identity = requirement.get("hashed_identity", "")
+        assert isinstance(bid_required_hashed_identity, str), \
+            f"Invalid requirement {requirement}"
+        offer_hashed_identity = (
+                offer.attributes or {}).get("hashed_identity", None)
+        if bid_required_hashed_identity and offer_hashed_identity != bid_required_hashed_identity:
+            return False
+        return True
```

Comment:
Double negations are always harder to understand IMO. How about:
```suggestion
        return True if offer_hashed_identity == bid_required_hashed_identity else False
```

Reply:
Same for the other Requirement classes.